### ISSUE TITLE
On neovim decompile does not work

### ISSUE THREAD

**Issue by Decodetalkers** [razzmatazz/csharp-language-server#21]
When I use csharpls, and use the jump of nvimlsp, it do not return a decompile file.

I see that you have support decompile, I think it should work

**Comment by razzmatazz** [comment]
hi @chen244 

support for decompilation is an extension of lsp protocol and needs work on the client to be supported

if you're willing to do this I can provide pointers; you should probably check java/clojure or other related nvimlsp client where decompilation is supported and I can help you hack it in

**Comment by chen244** [comment]
> hi @chen244
> 
> support for decompilation is an extension of lsp protocol and needs work on the client to be supported (works on emacs only for now, AFAIK)
> 
> if you're willing to do this I can provide pointers; you should probably check java/clojure or other related nvimlsp client where decompilation is supported, copy the code and I can help you hack it in

Thanks. what should I do , how to make a lsp extension to support decompile? I will try. Can you help me?

**Comment by razzmatazz** [comment]
can you find an existing nvimlsp client that suports decompilation? so we can have some starting base. maybe java or clojure or some other?

**Comment by chen244** [comment]
> can you find an existing nvimlsp client that suports decompilation? so we can have some starting base. maybe java or clojure or some other?

https://github.com/Hoffs/omnisharp-extended-lsp.nvim/issues/6 
But it is the autor says it is because the bug of omnisharp..

**Comment by razzmatazz** [comment]
this is a great start!

I have no experience in nvim/lua but copying most of the logic plus doing a couple of changes should work

**Comment by chen244** [comment]
> 

Thanks very much

**Comment by chen244** [comment]
```lua
local result, err = client.request_sync("o#/metadata", params, 10000)
```
in omnisharp , it get the decompile from o#/metadata, but how can I get the result in csharp_ls

**Comment by razzmatazz** [comment]
the handler for this on csharp-ls is "csharp/metadata":
- https://github.com/razzmatazz/csharp-language-server/blob/master/src/CSharpLanguageServer/Server.fs#L1206

you need to send this:
- https://github.com/razzmatazz/csharp-language-server/blob/master/src/CSharpLanguageServer/Server.fs#L72

and the response is:
- https://github.com/razzmatazz/csharp-language-server/blob/master/src/CSharpLanguageServer/Server.fs#L76

you can see emacs lsp-mode implementation here:
- https://github.com/emacs-lsp/lsp-mode/blob/master/clients/lsp-csharp.el#L381

please ask if you need request examples, I may be able to provide lsp traces for what emacs sends and receives so you can emulate this

**Comment by chen244** [comment]
> the handler for this on csharp-ls is `"csharp/metadata"`:
> 
>     * https://github.com/razzmatazz/csharp-language-server/blob/master/src/CSharpLanguageServer/Server.fs#L1206
> 
> 
> you need to send this:
> 
>     * https://github.com/razzmatazz/csharp-language-server/blob/master/src/CSharpLanguageServer/Server.fs#L72
> 
> 
> and the response is:
> 
>     * https://github.com/razzmatazz/csharp-language-server/blob/master/src/CSharpLanguageServer/Server.fs#L76
> 
> 
> you can see emacs lsp-mode implementation here:
> 
>     * https://github.com/emacs-lsp/lsp-mode/blob/master/clients/lsp-csharp.el#L381
> 
> 
> please ask if you need request examples, I may be able to provide lsp traces for what emacs sends and receives so you can emulate this

For example ,I get 
csharp:/metadata/projects/trainning2/assemblies/System.Console/symbols/System.Console.cs
this url

and I will get 
```lua
{
      assymblyName = "System.Console", 
     projectName = "trainning2",
      typeName = "System.Console"
}
```

Then I send this to csharp/metadata. is it right?

**Comment by razzmatazz** [comment]
emacs trace for go-to-def to `string`:

> <img width="484" alt="Screenshot 2022-03-08 at 13 15 02" src="https://user-images.githubusercontent.com/563556/157227841-c14fe3e8-1c9b-4b74-ae20-a6fd571a21ce.png">

resolving definition to csharp:/ url this is sent and received from the server:

> <img width="902" alt="Screenshot 2022-03-08 at 13 17 28" src="https://user-images.githubusercontent.com/563556/157228188-b2da9b62-04a1-45f7-b78c-85bb1d77a891.png">

then it sees that the url is `csharp:/` and retrieves metadata:

> <img width="1379" alt="Screenshot 2022-03-08 at 13 18 17" src="https://user-images.githubusercontent.com/563556/157228310-553a3418-9823-4963-b04a-133c1bb8fa45.png">

**Comment by chen244** [comment]
> emacs trace for go-to-def to `string`:
> 
> > <img alt="Screenshot 2022-03-08 at 13 15 02" width="484" src="https://user-images.githubusercontent.com/563556/157227841-c14fe3e8-1c9b-4b74-ae20-a6fd571a21ce.png">
> 
> resolving definition:
> 
> > <img alt="Screenshot 2022-03-08 at 13 17 28" width="902" src="https://user-images.githubusercontent.com/563556/157228188-b2da9b62-04a1-45f7-b78c-85bb1d77a891.png">
> 
> then it sees that the url is `csharp:/` and retrieves metadata:
> 
> > <img alt="Screenshot 2022-03-08 at 13 18 17" width="1379" src="https://user-images.githubusercontent.com/563556/157228310-553a3418-9823-4963-b04a-133c1bb8fa45.png">

I Got it! will

**Comment by razzmatazz** [comment]
great!

**Comment by razzmatazz** [comment]
can you add PR with FAQ entry to README.md on this project on how to make it work?

**Comment by chen244** [comment]
> 

Ok , I see

**Comment by chen244** [comment]
> btw, could you add PR with FAQ entry to README.md on this project on how to make it work?

Ok , I see. btw, you means your project, the csharp-language-server?

**Comment by razzmatazz** [comment]
> > btw, could you add PR with FAQ entry to README.md on this project on how to make it work?
> 
> Ok , I see. btw, you means your project, the csharp-language-server?

yes, here https://github.com/razzmatazz/csharp-language-server/blob/master/README.md

I see there is not FAQ yet, so it would be great if you could just drop a line to your project or general directions on to make metadata work on nvim
